vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -1,1011 +1,527 @@
+        mors = [m for m in mors if isinstance(m, resource_type)]
+
+        if resource_type == vim.ClusterComputeResource:
+            # Cluster metrics are unpredictable and a single call can max out the limit. Always collect them one by one.
+            max_batch_size = 1
+        elif resource_type in REALTIME_RESOURCES or self.max_historical_metrics < 0:
+            # Queries are not limited by vCenter
+            max_batch_size = self.metrics_per_query
+        else:
+            # Collection is limited by the value of `max_query_metrics` (aliased to self.max_historical_metrics)
+            if self.metrics_per_query < 0:
+                max_batch_size = self.max_historical_metrics
+            else:
+                max_batch_size = min(self.metrics_per_query, self.max_historical_metrics)
+
+        batch = defaultdict(list)
+        batch_size = 0
+        for m in mors:
+            for metric in metric_ids:
+                if batch_size == max_batch_size:
+                    yield batch
+                    batch = defaultdict(list)
+                    batch_size = 0
+                batch[m].append(metric)
+                batch_size += 1
+        # Do not yield an empty batch
+        if batch:
+            yield batch
+
+    def submit_external_host_tags(self):
+        """Send external host tags to the Datadog backend. This is only useful for a REALTIME instance because
+        only VMs and Hosts appear as 'datadog hosts'."""
+        external_host_tags = []
+        hosts = self.infrastructure_cache.get_mors(vim.HostSystem)
+        vms = self.infrastructure_cache.get_mors(vim.VirtualMachine)
+
+        for mor in chain(hosts, vms):
+            # Safeguard if some mors have a None hostname
+            mor_props = self.infrastructure_cache.get_mor_props(mor)
+            hostname = mor_props.get('hostname')
+            if not hostname:
+                continue
 
-        self.log.debug("Collecting metrics for %s mors", ensure_unicode(n_mors))
-
-        # Request metrics for several objects at once. We can limit the number of objects with batch_size
-        # If batch_size is 0, process everything at once
-        batch_size = self.batch_morlist_size or n_mors
-        for batch in mors_batch_method(i_key, batch_size, max_historical_metrics):
-            query_specs = []
-            for mor in itervalues(batch):
-                if mor['mor_type'] == 'vm':
-                    vm_count += 1
-                if mor['mor_type'] not in REALTIME_RESOURCES and ('metrics' not in mor or not mor['metrics']):
-                    continue
+            tags = [t for t in mor_props['tags'] if t.split(':')[0] not in self.excluded_host_tags]
+            tags.extend(self.base_tags)
+            external_host_tags.append((hostname, {self.__NAMESPACE__: tags}))
 
-                query_spec = vim.PerformanceManager.QuerySpec()
-                query_spec.entity = mor["mor"]
-   
```

Comment:
Maybe I don't understand this 100% (also the code linked to your comment seems unrelated)

It may indeed not be needed by generating all the data in one go but this is an important safeguard for the future as well. It also is consistant with the infrastructure cache pattern.

Reply:
I mean by using a code like this (a bit simpler IMO), you can avoid the use of the context manager.

```python
    # VSphereConfig
    def refresh_metrics_metadata_cache(self):
        ...
        metadata_dict = {}
        for mor_type in self.collected_resource_types:
            ....
            metadata_dict[mor_type] = metadata
        self.metrics_metadata_cache.set_all_metadata(metadata)

    # VSphereCache
    def set_all_metadata(self, metadata):
        self._content[resource_type] = metadata
        self._last_ts = time.time()
```